### DISCUSSION TITLE
How do i rebind multiple pci devices using the new machine config document PCIDriverRebindConfig ?

### DISCUSSION THREAD

**Discussion by d4rkfella** [siderolabs/talos#10025] [General]
Hello, i've been testing the new machine config document for rebinding pci devices to vfio driver, but i was wondering is it possible to rebind multiple devices ? I tried adding separate documents for each id but it wouldnt allow to have duplicates , i also tried naming them but that isnt allowed aswell.

**Comment by frezbo** [comment]
Could you provide an example use? Multiple named documents should be supported

**Comment by d4rkfella** [comment]
When i try to name the document i receive a validation error 
```  GNU nano 8.2                                                                                               /tmp/MachineConfigs.config.talos.dev-v1alpha1-edit-2365456790.yaml                                                                                                         
# Editing MachineConfigs.config.talos.dev/v1alpha1 at node 192.168.91.9
#
# 1 error occurred:
#       * 192.168.91.9: rpc error: code = InvalidArgument desc = unknown keys found during decoding:
# name: bind1
#
#
#
---
apiVersion: v1alpha1
kind: PCIDriverRebindConfig
name: bind1
pciID: 0000:0a:00.0
targetDriver: vfio-pci
---

**Reply by smira** [reply]
I'm not sure where does the `name` come from? Is it somewhere in the docs?

It should be `pciID`: https://www.talos.dev/v1.10/reference/configuration/hardware/pcidriverrebindconfig/

**Comment by d4rkfella** [comment]
Im trying to rebind 2 different pci devices to vfio driver and so i tried creating 2 separate documents and naming them, since unnamed documents can only be supplied once in the machine config. I was using this documentation as a reference https://www.talos.dev/v1.9/reference/configuration/

**Reply by frezbo** [reply]
the 1.9 docs shouldn't have that, I probably missed removing that

**Reply by d4rkfella** [reply]
Alright, what would be the correct way to provide multiple pciIDs for rebinding ?

**Reply by smira** [reply]
> the 1.9 docs shouldn't have that, I probably missed removing that

let's please remove that, I think I left a comment on your PR, Noel, to revert that

**Reply by smira** [reply]
> Alright, what would be the correct way to provide multiple pciIDs for rebinding ?

https://www.talos.dev/v1.10/reference/configuration/hardware/pcidriverrebindconfig/

It is only available in v1.10+ (not 1.9).

**Reply by d4rkfella** [reply]
Yea, i know that. I've already upgraded to the alpha version of 1.10. Just wondering how to set up multiple pci devices for rebinding.

**Reply by d4rkfella** [reply]
I tried creating 2 separate documents for each device and also providing each device in a single document, separating them by comma , but neither did work.

**Reply by smira** [reply]
We can't help unless you provide an exact machine configuration patch and the steps you did.

**Reply by d4rkfella** [reply]
Well i basically edited my machineconfig and added the driverrebind document to it like this .
```
apiVersion: v1alpha1
kind: PCIDriverRebindConfig
pciID: 0000:0a:00.0
targetDriver: vfio-pci
---
version: v1alpha1
debug: false
persist: true
machine
  <rest of the machineconfig>
```

**Reply by smira** [reply]
This is hard to digest, sorry.

**Reply by d4rkfella** [reply]
In what way ? You asked me to provide my machineconfig so i did that.

**Reply by d4rkfella** [reply]
I added the pcirebindconfig document directly by editing my machineconfig and separating it with dashes. It works perfectly for that single device, i just cant figure out how to add another device that i want to rebind to vfio aswell, if its even possible at this point ?

**Reply by d4rkfella** [reply]
Maybe i shouldnt have added the whole config , you are right. I cleaned it up.

**Reply by frezbo** [reply]
> Well i basically edited my machineconfig and added the driverrebind document to it like this .
> 
> ```
> apiVersion: v1alpha1
> kind: PCIDriverRebindConfig
> pciID: 0000:0a:00.0
> targetDriver: vfio-pci
> ---
> version: v1alpha1
> debug: false
> persist: true
> machine
>   <rest of the machineconfig>
> ```

This should work, you can add a second document like this

**Reply by d4rkfella** [reply]
I tried that but im receiving an error that duplicate document is not allowed. 
```# Editing MachineConfigs.config.talos.dev/v1alpha1 at node 192.168.91.9
#
# 1 error occurred:
#       * 192.168.91.9: rpc error: code = InvalidArgument desc = duplicate document v1alpha1/PCIDriverRebindConfig/ is not allowed

**Reply by frezbo** [reply]
thanks, I found the bug, will fix it for 1.9.1

**Reply by frezbo** [reply]
> thanks, I found the bug, will fix it for 1.9.1

sorry, 1.10.0-alpha.1, since this is 1.10 feature only

**Reply by d4rkfella** [reply]
Yea, i figured. Thanks for developing this feature and wishing you guys happy holidays .
